<a href="https://colab.research.google.com/github/ArvindRe/GuidedCapstone/blob/master/GuidedCapstoneStep4_and_step5HL.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Guided Capstone Step 4. Pre-Processing and Training Data Development

**The Data Science Method**  


1.   Problem Identification 


2.   Data Wrangling 
  
 
3.   Exploratory Data Analysis   

4.   **Pre-processing and Training Data Development**  
 * Create dummy or indicator features for categorical variables
  * Standardize the magnitude of numeric features
  * Split into testing and training datasets
  * Apply scaler to the testing set
5.   Modeling 
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

**<font color='teal'> Start by loading the necessary packages as we did in step 3 and printing out our current working directory just to confirm we are in the correct project directory. </font>**

In [137]:
#load python packages
import os
import pandas as pd
import datetime
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
import pandas.util.testing as tm
%matplotlib inline

**<font color='teal'>  Load the csv file you created in step 3, remember it should be saved inside your data subfolder and print the first five rows.</font>**

In [138]:
from google.colab import files
uploaded=files.upload()
for fn in uploaded.keys():
  print('User uploaded file "{name}" with length {length} bytes'.format(
      name=fn, length=len(uploaded[fn])))
import io
df=pd.read_csv(io.StringIO(uploaded['step3_output.csv'].decode('utf-8')))
df.head()

Saving step3_output.csv to step3_output (3).csv
User uploaded file "step3_output.csv" with length 22930 bytes


Unnamed: 0                           Name  ... NightSkiing_ac  clusters
0           2               Hilltop Ski Area  ...           30.0         2
1           4            Sunrise Park Resort  ...           80.0         1
2           5  Yosemite Ski & Snowboard Area  ...            0.0         1
3           8         Boreal Mountain Resort  ...          200.0         1
4           9                    Dodge Ridge  ...            0.0         1

[5 rows x 27 columns]

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `state`. Add the dummies back to the dataframe and remove the original column for `state`. </font>**

Hint: you can see an example of how to execute this in Aiden's article on preprocessing [here](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-pre-processing-and-training-data-development-fd2d75182967). 

In [139]:
df.iloc[:,2:]

state  summit_elev  ...  NightSkiing_ac  clusters
0        Alaska         2090  ...            30.0         2
1       Arizona        11100  ...            80.0         1
2    California         7800  ...             0.0         1
3    California         7700  ...           200.0         1
4    California         8200  ...             0.0         1
..          ...          ...  ...             ...       ...
168     Wyoming         8000  ...             0.0         1
169     Wyoming         7428  ...             0.0         1
170     Wyoming         7808  ...           110.0         1
171     Wyoming         9663  ...             0.0         1
172     Wyoming         9500  ...             0.0         1

[173 rows x 25 columns]

In [140]:
dfo=df.iloc[:,2:].select_dtypes(include=['object']) # select object type columns
df = pd.concat([df.drop(dfo, axis=1), pd.get_dummies(dfo)], axis=1)
#dfo.head()
df.head()


Unnamed: 0                           Name  ...  state_Wisconsin  state_Wyoming
0           2               Hilltop Ski Area  ...                0              0
1           4            Sunrise Park Resort  ...                0              0
2           5  Yosemite Ski & Snowboard Area  ...                0              0
3           8         Boreal Mountain Resort  ...                0              0
4           9                    Dodge Ridge  ...                0              0

[5 rows x 61 columns]

In [141]:
#df.drop(['Name'], axis=1,inplace=True)
#df.head()

## Standardize the magnitude of numeric features

In [142]:
#**<font color='teal'> Using sklearn preprocessing standardize the scale of the features of the dataframe except the name of the resort which we done't need in the dataframe for modeling, so it can be droppped here as well. Also, we want to hold out our response variable(s) so we can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as our response for scaling and modeling. Later we will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave them in the development dataframe. </font>**
#from sklearn import preprocessing
#df_normalized = preprocessing.normalize(df)
#df_standardized = preprocessing.scale(df)

In [143]:
#(df_normalized)

In [144]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = df.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df.loc[:,'AdultWeekend']

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)  #?

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X)  #?

## Split into training and testing datasets

**<font color='teal'> Using sklearn model selection import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. We will start by using the adult weekend ticket price as our response variable for modeling.</font>**

In [145]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

Here we start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5. Modeling


This is the fifth step in the Data Science Method. In the previous steps you cleaned and prepared the datasets. Now it's time to get into the most exciting part: modeling! In this exercise, you'll build three different models and compare each model's performance. In the end, you'll choose the best model for demonstrating insights to Big Mountain management.



### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with a Training Dataset

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1

In [146]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

Hint: you will have to use the `predict()` method here as it's used in this [DSM article](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-modeling-56b4233cad1b) about modeling. 

In [147]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred=model.predict(X_test)
print(y_pred)

[7.55037842e+01 6.51522217e+01 6.39881592e+01 6.20936279e+01
 5.13436279e+01 3.07484114e+13 6.98748779e+01 7.85311279e+01
 4.84061279e+01 5.45506592e+01 7.76561279e+01 8.05076904e+01
 6.36405029e+01 3.14664819e+13 3.33846436e+01 6.20067139e+01
 4.43123779e+01 4.57225342e+01 5.30194092e+01 3.42967529e+01
 5.92069092e+01 3.07280109e+13 6.39813232e+01 4.91873779e+01
 5.08045654e+01 4.73045654e+01 5.10623779e+01 4.77811279e+01
 7.49334717e+01 4.16766357e+01 4.92108154e+01 7.43485107e+01
 8.58670654e+01 7.14295654e+01 5.21678467e+01 5.45623779e+01
 3.44061279e+01 7.02498779e+01 4.74373779e+01 3.85858154e+01
 3.29686279e+01 4.39139404e+01 4.67030029e+01 5.04647217e+01]


## Review Model Outcomes — Iterate over additional models as needed

In [148]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from sklearn.metrics. 
# You can plug y_test and y_pred into the functions to evaluate the model
from sklearn.metrics import mean_absolute_error
from sklearn.metrics import explained_variance_score
mean_absolute_error=mean_absolute_error(y_test, y_pred)
explained_variance_score=explained_variance_score(y_test, y_pred)

In [149]:
print(explained_variance_score)

-2.1042265877418214e+23


In [150]:
print(mean_absolute_error)

2112338730992.0132


**<font color='teal'> Print the intercept value from the linear model. </font>**

Hint: our linear regression model `lm` has an attribute `intercept_` for the intercept

In [151]:
print('intercept:', lm.intercept_)

intercept: 537242220654.5829


**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the resulting `AdultWeekend` value. Also, because we took the time to scale our x values in the training data, we can compare each of the coeeficients for the features to determine the feature importances. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 


Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what we are looking for is the magnitude of impact on our response variable.

In [159]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
# pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])
df_coef=pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient']).sort_values(by='Coefficient',ascending=False)
df_coef

Coefficient
trams                 8.347773e+13
total_chairs          7.250868e+13
fastEight             4.744331e+13
double                3.876110e+13
surface               3.795128e+13
triple                3.236080e+13
quad                  1.843571e+13
fastQuads             1.539578e+13
state_New York        9.831896e+12
state_Michigan        8.652682e+12
state_California      7.812322e+12
state_Wisconsin       7.502922e+12
state_Pennsylvania    7.502922e+12
state_New Hampshire   7.502922e+12
state_Vermont         6.457487e+12
state_Minnesota       6.457487e+12
state_New Mexico      5.626139e+12
state_Idaho           5.626139e+12
state_Wyoming         5.151293e+12
state_Connecticut     5.151293e+12
state_Maine           5.151293e+12
state_Ohio            4.621148e+12
state_Colorado        4.621148e+12
state_Massachusetts   4.621148e+12
state_Washington      4.013855e+12
state_Utah            4.013855e+12
state_Oregon          4.013855e+12
state_Montana         4.013855e+12
state_West Virginia   4.013855e+12
state_Illinois        4.013855e+12
state_Iowa            4.013855e+12
state_North Carolina  4.013855e+12
state_Indiana         3.286924e+12
state_Missouri        3.286924e+12
state_Virginia        3.286924e+12
fastSixes             2.588798e+12
state_South Dakota    2.330992e+12
state_Arizona         2.330992e+12
state_Alaska          2.330992e+12
state_Tennessee       2.330992e+12
state_New Jersey      2.330992e+12
state_Maryland        5.443591e+10
state_Nevada          1.546499e+09
state_Rhode Island    3.888500e+04
AdultWeekday          8.031732e+00
Unnamed: 0            4.455656e+00
clusters              1.815979e+00
averageSnowfall       1.699293e+00
vertical_drop         1.234680e+00
Runs                  1.219630e+00
projectedDaysOpen     1.112246e+00
yearsOpen             1.087483e+00
NightSkiing_ac        1.041069e+00
LongestRun_mi         9.620882e-01
daysOpenLastYear      8.838055e-01
SkiableTerrain_ac     6.806452e-01
Snow Making_ac        4.761307e-01
TerrainParks          3.197366e-01
summit_elev           2.388893e-01

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that we care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

Hint: Try to construct another model using exactly the steps we followed above. 

In [153]:
def RegressionModel(X,y, test_size, random_state):
    try:
        # Test Train split
        X_train, X_test, y_train , y_test = train_test_split(X, y, test_size=test_size, random_state=random_state)
        # Fit the values into the model
        lm = linear_model.LinearRegression()
        model = lm.fit(X_train,y_train)
        # Predict model with test set
        y_pred = model.predict(X_test)
        
        explained_variance = explained_variance_score(y_test, y_pred)
        meanabsolute_error = mean_absolute_error(y_test, y_pred)
        intercept = lm.intercept_
        
        Coefficient = pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient']).sort_values(by='Coefficient',ascending=False)
        
        results = {'Explained Variance' : explained_variance,
                    'Mean absolute error': meanabsolute_error,
                  'intercept': intercept,
                  'Co-efficient': Coefficient}
        return results
    except Exception as e:
        return str(e)

In [154]:
from sklearn.ensemble import RandomForestRegressor
def RandomforestregressorModel(X,y, random_state, estimators):
  #spilt
  X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=1)
  #model
  # Fit the values into the model
  regressor = RandomForestRegressor(random_state=random_state, n_estimators=estimators)
  model2=regressor.fit(X_train, y_train)
  # Score and Predict model with test set
  model2_score=regressor.score(X_test, y_test)
  y_pred=model2.predict(X_test)
  
  #ev=explained_variance_score(y_test, y_pred)              #'numpy.float64' object is not callable?
  #mae=mean_absolute_error(y_test, y_pred)                  #'numpy.float64' object is not callable

  Coefficient = pd.DataFrame(model2.feature_importances_, X.columns, columns=['Coefficient']).sort_values(by='Coefficient',ascending=False)   #Coefficient vs importance?

  results = {'score' : model2_score,
            # 'Explained Variance' : ev,
            # 'Mean absolute error': mae,
             'Co-efficient': Coefficient}

  return results

#### Model 2

In [155]:
df.columns

Index(['Unnamed: 0', 'Name', 'summit_elev', 'vertical_drop', 'trams',
       'fastEight', 'fastSixes', 'fastQuads', 'quad', 'triple', 'double',
       'surface', 'total_chairs', 'Runs', 'TerrainParks', 'LongestRun_mi',
       'SkiableTerrain_ac', 'Snow Making_ac', 'daysOpenLastYear', 'yearsOpen',
       'averageSnowfall', 'AdultWeekday', 'AdultWeekend', 'projectedDaysOpen',
       'NightSkiing_ac', 'clusters', 'state_Alaska', 'state_Arizona',
       'state_California', 'state_Colorado', 'state_Connecticut',
       'state_Idaho', 'state_Illinois', 'state_Indiana', 'state_Iowa',
       'state_Maine', 'state_Maryland', 'state_Massachusetts',
       'state_Michigan', 'state_Minnesota', 'state_Missouri', 'state_Montana',
       'state_Nevada', 'state_New Hampshire', 'state_New Jersey',
       'state_New Mexico', 'state_New York', 'state_North Carolina',
       'state_Ohio', 'state_Oregon', 'state_Pennsylvania',
       'state_Rhode Island', 'state_South Dakota', 'state_Tennessee',
       'st

In [156]:
X_m2=df.drop(['Name','AdultWeekend','state_Alaska', 'state_Arizona',
       'state_California', 'state_Colorado', 'state_Connecticut',
       'state_Idaho', 'state_Illinois', 'state_Indiana', 'state_Iowa',
       'state_Maine', 'state_Maryland', 'state_Massachusetts',
       'state_Michigan', 'state_Minnesota', 'state_Missouri', 'state_Montana',
       'state_Nevada', 'state_New Hampshire', 'state_New Jersey',
       'state_New Mexico', 'state_New York', 'state_North Carolina',
       'state_Ohio', 'state_Oregon', 'state_Pennsylvania',
       'state_Rhode Island', 'state_South Dakota', 'state_Tennessee',
       'state_Utah', 'state_Vermont', 'state_Virginia', 'state_Washington',
       'state_West Virginia', 'state_Wisconsin', 'state_Wyoming'],axis=1)

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler1 = preprocessing.StandardScaler().fit(X_m2)  #?

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_m2_scaled=scaler1.transform(X_m2)  #?

In [157]:
#RandomforestregressorModel(X_m2_scaled,y, 0, 200) #scaling may not be required for randomforest?
RegressionModel(X_m2,y, 0.25, 1)

"'numpy.float64' object is not callable"

In [158]:
#X_train, X_test, y_train , y_test = train_test_split(X, y, test_size=0.25, random_state=1)
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)
        # Fit the values into the model
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)
        # Predict model with test set
y_pred = model.predict(X_test)
        
explained_variance = explained_variance_score(y_test, y_pred)
meanabsolute_error = mean_absolute_error(y_test, y_pred)
intercept = lm.intercept_
        
Coefficient = pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient']).sort_values(by='Coefficient',ascending=False)
        
results = {'Explained Variance' : explained_variance,
                    'Mean absolute error': meanabsolute_error,
                  'intercept': intercept,
                  'Co-efficient': Coefficient}
prin(results)

TypeError: ignored

In [160]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = df.drop(['Name','AdultWeekend','state_Alaska', 'state_Arizona',
       'state_California', 'state_Colorado', 'state_Connecticut',
       'state_Idaho', 'state_Illinois', 'state_Indiana', 'state_Iowa',
       'state_Maine', 'state_Maryland', 'state_Massachusetts',
       'state_Michigan', 'state_Minnesota', 'state_Missouri', 'state_Montana',
       'state_Nevada', 'state_New Hampshire', 'state_New Jersey',
       'state_New Mexico', 'state_New York', 'state_North Carolina',
       'state_Ohio', 'state_Oregon', 'state_Pennsylvania',
       'state_Rhode Island', 'state_South Dakota', 'state_Tennessee',
       'state_Utah', 'state_Vermont', 'state_Virginia', 'state_Washington',
       'state_West Virginia', 'state_Wisconsin', 'state_Wyoming'],axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df.loc[:,'AdultWeekend']

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)  #?

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X)  #?
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred=model.predict(X_test)
print(y_pred)
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from sklearn.metrics. 
# You can plug y_test and y_pred into the functions to evaluate the model
from sklearn.metrics import mean_absolute_error
from sklearn.metrics import explained_variance_score
mean_absolute_error=mean_absolute_error(y_test, y_pred)
explained_variance_score=explained_variance_score(y_test, y_pred)
print('explained_variance_score:',explained_variance_score)
print('mean_absolute_error:',mean_absolute_error)
print('intercept:', lm.intercept_)
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
# pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])
df_coef=pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient']).sort_values(by='Coefficient')
df_coef.T

[72.89372384 55.64197782 59.6724212  65.09170545 53.58012387 56.35867771
 65.85302275 71.3421087  50.04068906 56.70621393 78.56634416 76.6826614
 62.56646427 77.13974279 32.05618486 60.30764805 48.34219002 50.78241971
 52.34547326 47.56945412 60.21231197 62.42207057 61.10456419 54.72609311
 43.29953092 41.22561685 54.58773111 55.97335981 70.89341149 38.90990428
 52.35382027 75.33011283 92.86784169 68.47585072 57.27265298 49.98908162
 27.42885103 71.89948562 53.94093658 43.65344688 35.12545349 45.20803115
 50.58343402 41.15500844]
0.7361554989679906
6.941916025253085
intercept: 56.8047398792822


trams     fastSixes  ...  summit_elev  AdultWeekday
Coefficient  2.220446e-16  8.881784e-16  ...     3.746205      9.027718

[1 rows x 24 columns]

**<font color='teal'> When reviewing our new model coefficients, we see `summit_elev` is now in the number two spot. This is also difficult to change from a management prespective and highly correlated with `base_elev` and `vertical_drop`.  This time, rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

#### Model 3

In [161]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = df.drop(['Name','AdultWeekend','state_Alaska', 'state_Arizona',
       'state_California', 'state_Colorado', 'state_Connecticut',
       'state_Idaho', 'state_Illinois', 'state_Indiana', 'state_Iowa',
       'state_Maine', 'state_Maryland', 'state_Massachusetts',
       'state_Michigan', 'state_Minnesota', 'state_Missouri', 'state_Montana',
       'state_Nevada', 'state_New Hampshire', 'state_New Jersey',
       'state_New Mexico', 'state_New York', 'state_North Carolina',
       'state_Ohio', 'state_Oregon', 'state_Pennsylvania',
       'state_Rhode Island', 'state_South Dakota', 'state_Tennessee',
       'state_Utah', 'state_Vermont', 'state_Virginia', 'state_Washington',
       'state_West Virginia', 'state_Wisconsin', 'state_Wyoming','summit_elev'],axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df.loc[:,'AdultWeekend']

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)  #?

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X)  #?
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred=model.predict(X_test)
print(y_pred)
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from sklearn.metrics. 
# You can plug y_test and y_pred into the functions to evaluate the model
from sklearn.metrics import mean_absolute_error
from sklearn.metrics import explained_variance_score
mean_absolute_error=mean_absolute_error(y_test, y_pred)
explained_variance_score=explained_variance_score(y_test, y_pred)
print('explained_variance_score:',explained_variance_score)
print('mean_absolute_error:',mean_absolute_error)
print('intercept:', lm.intercept_)
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
# pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])
df_coef=pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient']).sort_values(by='Coefficient')
df_coef.T

[80.66135982 55.68681563 58.03636738 65.33470671 49.91442461 55.15608792
 66.98539323 73.69834307 49.69706934 56.10487149 82.09518736 74.9677997
 60.98082126 75.09239836 31.48147953 59.00973871 47.18605187 50.22679332
 53.91656483 44.8077404  58.15747961 69.94481718 60.96805277 52.98027266
 42.73960705 40.79206669 53.60627968 55.26730797 74.98999713 39.96112138
 52.53354426 76.0136239  85.57677341 68.590818   55.66299754 54.64947999
 34.14716257 69.58788645 53.57832792 43.76308804 32.79280795 44.15033412
 50.69320165 38.79612937]
0.7574119166688689
6.506166903763969
intercept: 56.797834349127164


trams     fastSixes  ...  vertical_drop  AdultWeekday
Coefficient    0.0  1.110223e-16  ...       2.403534      8.828939

[1 rows x 23 columns]

## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell — you will discuss this selection more in the next step of the guided casptone. </font>**

| Model | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1. | -2.1042265877418214e+23 | 2112338730992.0132 |-|
| Model 2. | 0.7361554989679906| 6.941916025253085 |'state'|
| Model 3. | 0.7574119166688689 | 6.506166903763969 |'state','summit_elev','base_elev'|

Model Selection:

Model Selection: Both model 2 & model 3 are nearly identical and giving almost the same variance and MAE. Although both models look the same we'll take model 3 as some features in the model 2 are not important to the problem statement.